### PR TITLE
[NEW] added code for download my data functionality

### PR DESCRIPTION
**Pull request by ShafiSingla1989**
@RocketChat/android

Closes #ISSUE_NUMBER

### CODE REVIEW

**Comment by bizzbyster** [comment]
@ShafiSingla1989 you should update your PR description to reference the issue, which is https://github.com/RocketChat/Rocket.Chat.Android/issues/1847.

File: app/src/main/java/chat/rocket/android/settings/presentation/SettingsView.kt [hunk 1]
```diff
@@ -1,3 +1,9 @@
 package chat.rocket.android.settings.presentation
 
-interface SettingsView
+import chat.rocket.android.core.behaviours.MessageView
+
+interface SettingsView:  MessageView {
+
+    fun showDownloadDialog(error : String?)
```
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/settings/presentation/SettingsView.kt]: Want to show an error message? You can use the function defined on `MessageView` instead of creating this one.

File: app/src/main/java/chat/rocket/android/settings/presentation/settingPresenter.kt [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    private val serverUrl = serverInteractor.get()!!
+    private val client: RocketChatClient = factory.create(serverUrl)
+
+
+
+
+
+    fun downloadData() {
+        launchUI(strategy) {
+            //view.showLoading()
```
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/settings/presentation/settingPresenter.kt]: The view should display a loading when executing a task.

File: app/src/main/java/chat/rocket/android/settings/presentation/settingPresenter.kt [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+            try {
+                withContext(DefaultDispatcher) {
+                    // REMARK: Backend API is only working with a lowercase hash.
+                    // https://github.com/RocketChat/Rocket.Chat/issues/12573
+
+                    val me = retryIO("me") { client.me() }
```
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/settings/presentation/settingPresenter.kt]: You don't need to invoke the `client.me()` to get the user id. You are using more background task doing that here.

File: app/src/main/java/chat/rocket/android/settings/presentation/settingPresenter.kt [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+
+
+
+    fun downloadData() {
+        launchUI(strategy) {
+            //view.showLoading()
+            try {
+                withContext(DefaultDispatcher) {
+                    // REMARK: Backend API is only working with a lowercase hash.
```
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/settings/presentation/settingPresenter.kt]: Where is this comment being related to the code here?

File: app/src/main/java/chat/rocket/android/settings/presentation/settingPresenter.kt [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+                    // https://github.com/RocketChat/Rocket.Chat/issues/12573
+
+                    val me = retryIO("me") { client.me() }
+                    val response = retryIO { client.requestDataDownload(me.id) }
+                   // view.hideLoading()
```
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/settings/presentation/settingPresenter.kt]: Why you added this comment here?

File: app/src/main/java/chat/rocket/android/settings/presentation/settingPresenter.kt [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+                    if(response.requested){
+                    view.showDownloadDialog("UserDataDownload_Requested_Text")}
```
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/settings/presentation/settingPresenter.kt]: Please remove this hard coded string.
This is also not meaningful.

File: app/src/main/java/chat/rocket/android/settings/presentation/settingPresenter.kt [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+                   else if(response.status.equals("completed")){
+                        view.showDownloadDialog("UserDataDownload_CompletedRequestExisted_Text")
```
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/settings/presentation/settingPresenter.kt]: Please remove this hard coded string.
This is also not meaningful.

File: app/src/main/java/chat/rocket/android/settings/presentation/settingPresenter.kt [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+                    }else
+                        view.showDownloadDialog("UserDataDownload_RequestExisted_Text")
```
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/settings/presentation/settingPresenter.kt]: Please remove this hard coded string.
This is also not meaningful.

File: app/src/main/java/chat/rocket/android/settings/presentation/settingPresenter.kt [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+                }
+            } catch (exception: Exception) {
+                exception.message?.let {
+                    view.showMessage(it)
+                }.ifNull {
+                    view.showGenericErrorMessage()
+                }
+            } finally {
+              //  view.hideLoading()
```
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/settings/presentation/settingPresenter.kt]: It should hide the loading when task is finalized.

File: app/src/main/java/chat/rocket/android/settings/ui/SettingsFragment.kt [hunk 10]
```diff
@@ -21,22 +21,26 @@ import chat.rocket.android.preferences.ui.PreferencesFragment
 import chat.rocket.android.preferences.ui.TAG_PREFERENCES_FRAGMENT
 import chat.rocket.android.settings.password.ui.PasswordActivity
 import chat.rocket.android.settings.presentation.SettingsView
-import chat.rocket.android.util.extensions.addFragmentBackStack
-import chat.rocket.android.util.extensions.inflate
+import chat.rocket.android.settings.presentation.settingPresenter
+import chat.rocket.android.util.extensions.*
```
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/settings/ui/SettingsFragment.kt:27]: Remove this wildcard import.

File: app/src/main/java/chat/rocket/android/settings/ui/SettingsFragment.kt [hunk 11]
```diff
@@ -116,4 +121,36 @@ class SettingsFragment : Fragment(), SettingsView, AdapterView.OnItemClickListen
     companion object {
         fun newInstance() = SettingsFragment()
     }
+
+    override fun showDownloadDialog(data : String?) {
+        ui {
+            val context = this
+            val builder = AlertDialog.Builder(activity)
+            val view = layoutInflater.inflate(R.layout.dialog_download, null)
+            if (data.equals("UserDataDownload_RequestExisted_Text"))
```
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/settings/ui/SettingsFragment.kt]: Not a big deal to declare this statement like this.

File: app/src/main/java/chat/rocket/android/settings/ui/SettingsFragment.kt [hunk 12]
```diff
@@ -116,4 +121,36 @@ class SettingsFragment : Fragment(), SettingsView, AdapterView.OnItemClickListen
     companion object {
         fun newInstance() = SettingsFragment()
     }
+
+    override fun showDownloadDialog(data : String?) {
+        ui {
+            val context = this
```
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/settings/ui/SettingsFragment.kt]: Where are you using this variable?

File: app/src/main/res/layout/dialog_download.xml [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+    android:padding="@dimen/padding_quote">
+
+    <TextView
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:text="@string/request_download"
+        android:layout_margin="@dimen/padding_quote"
+        android:layout_alignParentTop="true"
+        android:textColor="@color/color_black"
+        />
```
Comment (philipbrito) [inline app/src/main/res/layout/dialog_download.xml]: Bad formatting.

File: app/src/main/res/values/dimens.xml [hunk 14]
```diff
@@ -47,5 +47,7 @@
     <dimen name="viewer_toolbar_title">16sp</dimen>
 
     <dimen name="quote_collapsed_height">32dp</dimen>
+    <dimen name="rc_16dp">16dp</dimen>
```
Comment (philipbrito) [inline app/src/main/res/values/dimens.xml:50]: What `rc` stands for?

**Review by philipbrito — CHANGES_REQUESTED** [review]

File: app/src/main/res/values/dimens.xml (hunk 14 above)
Comment (ShafiSingla1989) [reply app/src/main/res/values/dimens.xml:50]: Rc stands for rocket chat. My idea is to give generic names to these dimensions. It could be resolved by naming it 16dp instead of rc_16dp.

File: app/src/main/java/chat/rocket/android/settings/presentation/settingPresenter.kt (hunk 2 above)
Comment (ShafiSingla1989) [reply app/src/main/java/chat/rocket/android/settings/presentation/settingPresenter.kt]: Message in the dialog depends on the response of task.

File: app/src/main/java/chat/rocket/android/settings/presentation/SettingsView.kt (hunk 1 above)
Comment (ShafiSingla1989) [reply app/src/main/java/chat/rocket/android/settings/presentation/SettingsView.kt]: sure

File: app/src/main/java/chat/rocket/android/settings/presentation/settingPresenter.kt (hunk 4 above)
Comment (ShafiSingla1989) [reply app/src/main/java/chat/rocket/android/settings/presentation/settingPresenter.kt]: will add a comment.

File: app/src/main/java/chat/rocket/android/settings/presentation/settingPresenter.kt (hunk 5 above)
Comment (ShafiSingla1989) [reply app/src/main/java/chat/rocket/android/settings/presentation/settingPresenter.kt]: will surely remove it

File: app/src/main/java/chat/rocket/android/settings/ui/SettingsFragment.kt (hunk 11 above)
Comment (ShafiSingla1989) [reply app/src/main/java/chat/rocket/android/settings/ui/SettingsFragment.kt]: Improved the code.

File: app/src/main/java/chat/rocket/android/settings/ui/SettingsFragment.kt (hunk 12 above)
Comment (ShafiSingla1989) [reply app/src/main/java/chat/rocket/android/settings/ui/SettingsFragment.kt]: Will remove this val.

File: app/src/main/res/layout/dialog_download.xml (hunk 13 above)
Comment (ShafiSingla1989) [reply app/src/main/res/layout/dialog_download.xml]: Can you please explain this in detail?

File: app/src/main/java/chat/rocket/android/settings/presentation/SettingsView.kt [hunk 15]
```diff
@@ -1,9 +1,10 @@
 package chat.rocket.android.settings.presentation
 
+import chat.rocket.android.core.behaviours.LoadingView
 import chat.rocket.android.core.behaviours.MessageView
 
-interface SettingsView:  MessageView {
+interface SettingsView:  MessageView, LoadingView {
```
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/settings/presentation/SettingsView.kt]: You don't need to add the brackets for an empty (no declarations) interface.

File: app/src/main/java/chat/rocket/android/settings/presentation/settingPresenter.kt [hunk 16]
```diff
@@ -27,25 +28,22 @@ class settingPresenter @Inject constructor(
 
 
 
-
+// this method is written to execute download data call
```
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/settings/presentation/settingPresenter.kt]: What I want to mean was: You don't need to leave the previous comment here (copy and paste error?) since there is no relation with your code.

The comment is ok now, but from the signature of the function it is meaningful enough. I mean: Not need to do that, but ok.

File: app/src/main/java/chat/rocket/android/settings/presentation/settingPresenter.kt [hunk 17]
```diff
@@ -27,25 +28,22 @@ class settingPresenter @Inject constructor(
...
-                    // REMARK: Backend API is only working with a lowercase hash.
-                    // https://github.com/RocketChat/Rocket.Chat/issues/12573
 
                     val me = retryIO("me") { client.me() }
                     val response = retryIO { client.requestDataDownload(me.id) }
-                   // view.hideLoading()
+                   view.hideLoading()
                     if(response.requested){
-                    view.showDownloadDialog("UserDataDownload_Requested_Text")}
+                    view.showMessage(R.string.UserDataDownload_Requested_Text)}
```
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/settings/presentation/settingPresenter.kt]: Bad declaration for the strings.

Please look on how we declare the stings from the corresponding files.

File: app/src/main/java/chat/rocket/android/settings/presentation/settingPresenter.kt [hunk 18]
```diff
@@ -27,25 +28,22 @@ class settingPresenter @Inject constructor(
...
                    else if(response.status.equals("completed")){
-                        view.showDownloadDialog("UserDataDownload_CompletedRequestExisted_Text")
+                        view.showMessage(R.string.UserDataDownload_CompletedRequestExisted_Text)
```
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/settings/presentation/settingPresenter.kt]: Bad declaration for the strings.

Please look on how we declare the stings from the corresponding files.

File: app/src/main/java/chat/rocket/android/settings/presentation/settingPresenter.kt [hunk 19]
```diff
@@ -27,25 +28,22 @@ class settingPresenter @Inject constructor(
...
                     }else
-                        view.showDownloadDialog("UserDataDownload_RequestExisted_Text")
-
+                        view.showMessage(R.string.UserDataDownload_RequestExisted_Text)
```
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/settings/presentation/settingPresenter.kt]: Bad declaration for the strings.

Please look on how we declare the stings from the corresponding files.

File: app/src/main/java/chat/rocket/android/settings/ui/SettingsFragment.kt [hunk 20]
```diff
@@ -25,28 +26,31 @@ import chat.rocket.android.settings.presentation.settingPresenter
...
+    override fun onCreate(savedInstanceState: Bundle?) {
+        AndroidSupportInjection.inject(this)
+        super.onCreate(savedInstanceState)
     }
 
     override fun onCreateView(
-        inflater: LayoutInflater,
-        container: ViewGroup?,
-        savedInstanceState: Bundle?
+            inflater: LayoutInflater,
```
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/settings/ui/SettingsFragment.kt:51]: Bad formatting.

File: app/src/main/java/chat/rocket/android/settings/ui/SettingsFragment.kt [hunk 21]
```diff
@@ -66,8 +70,8 @@ class SettingsFragment : Fragment(), SettingsView, AdapterView.OnItemClickListen
         when (parent?.getItemAtPosition(position).toString()) {
             resources.getString(R.string.title_preferences) -> {
                 (activity as AppCompatActivity).addFragmentBackStack(
-                    TAG_PREFERENCES_FRAGMENT,
-                    R.id.fragment_container
+                        TAG_PREFERENCES_FRAGMENT,
```
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/settings/ui/SettingsFragment.kt:73]: Bad formatting.

File: app/src/main/java/chat/rocket/android/settings/ui/SettingsFragment.kt [hunk 22]
```diff
@@ -76,8 +80,8 @@ class SettingsFragment : Fragment(), SettingsView, AdapterView.OnItemClickListen
                 startNewActivity(PasswordActivity::class)
             resources.getString(R.string.title_about) -> {
                 (activity as AppCompatActivity).addFragmentBackStack(
-                    TAG_ABOUT_FRAGMENT,
-                    R.id.fragment_container
+                        TAG_ABOUT_FRAGMENT,
```
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/settings/ui/SettingsFragment.kt:100]: Bad formatting.

File: app/src/main/java/chat/rocket/android/settings/ui/SettingsFragment.kt [hunk 23]
```diff
@@ -122,35 +126,35 @@ class SettingsFragment : Fragment(), SettingsView, AdapterView.OnItemClickListen
...
-            if (data.equals("UserDataDownload_RequestExisted_Text"))
-                view.text_download_description.textContent = resources.getString(R.string.UserDataDownload_RequestExisted_Text);
-            else if (data.equals("UserDataDownload_CompletedRequestExisted_Text"))
-                view.text_download_description.textContent = resources.getString(R.string.UserDataDownload_CompletedRequestExisted_Text);
-            else
-                view.text_download_description.textContent = resources.getString(R.string.UserDataDownload_Requested_Text);
 
+
+    override fun showMessage(resId: Int) {
+        ui { val builder = AlertDialog.Builder(activity)
```
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/settings/ui/SettingsFragment.kt:158]: Bad formatting.

File: app/src/main/java/chat/rocket/android/settings/ui/SettingsFragment.kt [hunk 24]
```diff
@@ -122,35 +126,35 @@ class SettingsFragment : Fragment(), SettingsView, AdapterView.OnItemClickListen
...
+            val view = layoutInflater.inflate(R.layout.dialog_download, null)
+            view.text_download_description.textContent = resources.getString(resId);
             builder.setView(view)
             builder.setPositiveButton(R.string.msg_ok) { dialog, p1 ->
```
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/settings/ui/SettingsFragment.kt]: `p1` is not being used. Please change to `_`.

File: app/src/main/res/layout/dialog_download.xml (hunk 13 above)
Comment (philipbrito) [reply app/src/main/res/layout/dialog_download.xml]: `android:textColor="@color/color_black" />` instead of
```
android:textColor="@color/color_black"
/>
```

**Review by philipbrito — CHANGES_REQUESTED** [review]
Please, take a look on how we declare the functions, structure the code and indent that.

File: app/src/main/java/chat/rocket/android/settings/ui/SettingsFragment.kt (hunk 24 above)
Comment (ShafiSingla1989) [reply app/src/main/java/chat/rocket/android/settings/ui/SettingsFragment.kt]: ok done

File: app/src/main/java/chat/rocket/android/settings/ui/SettingsFragment.kt (hunk 23 above)
Comment (ShafiSingla1989) [reply app/src/main/java/chat/rocket/android/settings/ui/SettingsFragment.kt:158]: What is bad formatting here

**Comment by rafaelks** [comment]
@ShafiSingla1989 Can you please take a look on the conflicts?

**Comment by ShafiSingla1989** [comment]
Hi have resolved the conflicts

**Comment by ShafiSingla1989** [comment]
I have resolved the conflict.

**Review by ShafiSingla1989 — COMMENTED** [review]
Hi i have resolved the conflits

**Comment by philipbrito** [comment]
As discussed before, we'll close this PR since it needs some improvements here (regarding design questions).

Thank you @ShafiSingla1989
